packages/lesswrong/server/resolvers/importUrlAsDraftPost.ts

```diff
@@ -0,0 +1,133 @@
+  }
+`);
+
+// Define WebsiteData GraphQL type
+addGraphQLSchema(`
+  type WebsiteData {
+    _id: String!
+    slug: String
+    title: String
+    url: String
+    postedAt: Date
+    createdAt: Date
+    userId: String
+    modifiedAt: Date
+    draft: Boolean
+    content: String
+    coauthorStatuses: [CoauthorStatus]
+  }
+`);
+
+// todo various url validation
+// mostly client side, but also mb avoid links to lw, eaf, etc
+export async function importUrlAsDraftPost(url: string, context: ResolverContext): Promise<WebsiteData> {
+  if (!context.currentUser) {
+    throw new Error('You must be logged in to fetch website HTML.')
+  }
+
+  const reviewUser = await Users.findOne({slug: 'review-bot'})
+  if (!reviewUser) {
+    throw new Error('Misconfiguration: Failed to find review user')
+  }
+
+    // todo what if I want to delete failed attempts and try again
+  // mb deletedDraft - false?
+  const existingPost = await fetchFragmentSingle({
+    collectionName: 'Posts',
+    fragmentName: 'PostsEditQueryFragment',
+    currentUser: context.currentUser,
+    selector: {url, userId: reviewUser._id},
+    context,
+  })
+
+  if (existingPost) {
+    const latestRevision = await getLatestContentsRevision(existingPost, context);
+    return { _id: existingPost._id, slug: existingPost.slug, title: existingPost.title, url: existingPost.url, postedAt: existingPost.postedAt, createdAt: existingPost.createdAt, userId: existingPost.userId, coauthorStatuses: existingPost.coauthorStatuses, draft: existingPost.draft, modifiedAt: existingPost.modifiedAt, content: sanitize(latestRevision?.html ?? '')} 
+  }
+
+  let extractedData
+  
+  if (url.toLowerCase().includes('arxiv.org')) {
+    const arxivData = await ArxivExtractor.extract(url)
+
+    // take the abstract and prepend the authors list and append a hyperlink to the original with the text "Read the full paper"
+    const annotatedContents = `<p><strong>By ${arxivData.authors.join(", ")}</strong></p>
+<h2>Abstract</h2>
+<blockquote>${arxivData.abstract}</blockquote>
+<p><a href="${arxivData.pdfUrl}">Read the full paper</a></p>`
+
+    extractedData = {
+      title: arxivData.title,
+      content: annotatedContents,
+      published: arxivData.published,
+    }
+  } else {
+    extractedData = await extract(url)
+    if (!extractedData) {
+      throw new Error('Failed to extract data from URL')
+    }
+  }
+
+  const {data } = await createMutator({
+    collection: Posts,
+    document: {
+      userId: reviewUser._id,
+      title: extractedData.title,
+      url: url,
+      contents: {originalContents: {data: extractedData.content, type: 'ckEditorMarkup'}},
+      postedAt: extractedData.published ? new Date(extractedData.published) : undefined,
+      modifiedAt: new Date(),
+      coauthorStatuses: [{userId: context.currentUser._id, confirmed: true, requested: true}],
+      hasCoauthorPermission: true,
+      draft: true,
+      contents_latest: sanitize(extractedData.content ?? ''),
```

Comment:
This is not what `contents_latest` is - `contents_latest` is the revision ID that points to the latest (published) revision, and should get filled in automatically by the make_editable_callbacks.  (Also sanitization should in theory happen there as well.)